pkg/cortex/runtime_config.go

```diff
@@ -45,17 +45,38 @@ func loadRuntimeConfig(r io.Reader) (interface{}, error) {
 	return overrides, nil
 }
 
+// tenantLimitsFromRuntimeConfig returns a function that translates tenant IDs to
+// specific limits for that tenant if configured, nil otherwise
 func tenantLimitsFromRuntimeConfig(c *runtimeconfig.Manager) validation.TenantLimits {
 	if c == nil {
 		return nil
 	}
+
+	supplier := tenantLimitsRuntimeConfigFunc(c)
 	return func(userID string) *validation.Limits {
-		cfg, ok := c.GetConfig().(*runtimeConfigValues)
-		if !ok || cfg == nil {
+		tenantLimits := supplier()
+		if tenantLimits == nil {
 			return nil
 		}
 
-		return cfg.TenantLimits[userID]
+		return tenantLimits[userID]
+	}
+}
+
+// tenantLimitsRuntimeConfigFunc returns a function that returns a mapping of all
```

Comment:
Using a function here like you mentioned and the code definitely seems simpler. Not sure if this is exactly what you had in mind @pstibrany